### PR TITLE
Add support of MTP

### PR DESCRIPTION
**Pull request by czhu15**
example code:
    import os

    os.environ['VLLM_SKIP_WARMUP'] = 'true'

    os.environ["VLLM_EP_SIZE"] = "8"
    os.environ["PT_HPU_ENABLE_LAZY_COLLECTIVES"] = "true"
    os.environ["PT_HPU_WEIGHT_SHARING"] = "0"
    os.environ["VLLM_MLA_PERFORM_MATRIX_ABSORPTION"]= "1"
    os.environ['VLLM_CONTIGUOUS_PA'] = 'false'
    os.environ['VLLM_MLA_DISABLE_REQUANTIZATION']='1'

    from vllm import LLM, SamplingParams

    prompts = [
        "Hello, my name is",
        "The president of the United States is",
        "The capital of France is",
        "The future of AI is",
    ]
    model_path="/local_dataset_2/ds_r1/DeepSeek-R1-fp8-G2/DeepSeek-R1-G2-static"

    if __name__ == "__main__":
        # Create a sampling params object.
        sampling_params = SamplingParams(temperature=0.0,
                                        max_tokens=16)

        llm = LLM(model=model_path,
                    trust_remote_code=True,
                    enforce_eager=True,
                    dtype="bfloat16",
                    use_v2_block_manager=True,
                    tensor_parallel_size=8,
                    max_model_len=1024,
                    num_scheduler_steps=1,
                    distributed_executor_backend='mp',
                    gpu_memory_utilization=0.8,
                    # kv_cache_dtype="fp8_inc",
                    enable_expert_parallel=True,
                    num_speculative_tokens=1,
                    max_num_seqs=4,
                    seed=2024)

        # Generate texts from the prompts. The output is a list of RequestOutput objects
        # that contain the prompt, generated text, and other information.
        outputs = llm.generate(prompts, sampling_params)
        # Print the outputs.
        for output in outputs:
            prompt = output.prompt
            generated_text = output.outputs[0].text
            print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
    output:
    Prompt: 'Hello, my name is', Generated text: '\n\n# Dr. Kaitlin Harkness\n\n## Professional information\n\n-'
    Prompt: 'The president of the United States is', Generated text: ' the head of state and head of government of the United States, indirectly elected to'
    Prompt: 'The capital of France is', Generated text: ' Paris. Paris is located in the northern part of the country, along the Seine'
    Prompt: 'The future of AI is', Generated text: " here, and it's called GPT-4. This groundbreaking language model is set"

### CODE REVIEW

File: vllm/attention/layer.py [hunk 1]
```diff
@@ -164,17 +166,25 @@ def forward(
...
+            if not self.use_mla:
+                # Reshape the query, key, and value tensors.
+                # NOTE(woosuk): We do this outside the custom op to minimize the
+                # CPU overheads from the non-CUDA-graph regions.
+                query = query.view(-1, self.num_heads, self.head_size)
+                output = output.view(-1, self.num_heads, self.head_size)
+                if key is not None:
+                    key = key.view(-1, self.num_kv_heads, self.head_size)
+                if value is not None:
+                    value = value.view(-1, self.num_kv_heads, self.head_size)
```
Comment (xuechendi) [inline vllm/attention/layer.py:187]: Is this change from upstream MTP changes?

File: vllm/model_executor/models/deepseek_mtp.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        previous_hidden_states = self.hnorm(previous_hidden_states)
+
+        hidden_states = self.eh_proj(
+            torch.cat([inputs_embeds, previous_hidden_states], dim=-1))
+
+        hidden_states, residual = self.mtp_block(positions=positions,
+                                                 hidden_states=hidden_states,
+                                                 residual=None)
+        hidden_states = residual + hidden_states
+        return self.shared_head(hidden_states)
```
Comment (xuechendi) [inline vllm/model_executor/models/deepseek_mtp.py:91]: suggest to split this PR into multiple commits, firstly only bring in upstream commit, then apply changes for HPU

File: vllm/model_executor/models/deepseek_v2.py [hunk 3]
```diff
@@ -500,8 +498,6 @@ def forward(
         self,
         positions: torch.Tensor,
         hidden_states: torch.Tensor,
-        kv_cache: torch.Tensor,
-        attn_metadata: AttentionMetadata,
```
Comment (xuechendi) [inline vllm/model_executor/models/deepseek_v2.py:504]: for updating deepseek_v1 by passing kv_cache and attn_metadata using forward_context, suggest to use another PR to do the update. I assume this is not related to MTP?

File: vllm/model_executor/models/interfaces_base.py [hunk 4]
```diff
@@ -62,7 +59,7 @@ def _check_vllm_model_forward(model: Union[Type[object], object]) -> bool:
     if not callable(model_forward):
         return False
 
-    vllm_kws = ("input_ids", "positions", "kv_caches", "attn_metadata")
```
Comment (xuechendi) [inline vllm/model_executor/models/interfaces_base.py:65]: same suggestion here, it will be better to use seperate PR or commit to do the API change

File: vllm/attention/layer.py (hunk 1 above)
Comment (czhu15) [reply vllm/attention/layer.py:187]: yes. this is the implementation from deepseek_r1_upstream branch

File: vllm/model_executor/models/deepseek_mtp.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+    def compute_logits(
+        self,
+        hidden_states: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        spec_step_idx: int = 0,
+    ) -> torch.Tensor:
+        mtp_layer = self.layers[str(self.mtp_start_layer_idx + spec_step_idx)]
+        logits = self.logits_processor(mtp_layer.shared_head.head,
+                                       hidden_states, sampling_metadata)
```
[the remark this answers is not in the document]
Comment (czhu15) [reply vllm/model_executor/models/deepseek_mtp.py:141]: Currently I use the deepseek_mtp.py version from the first commit in the community.
https://github.com/HabanaAI/vllm-fork/commit/f525c0be8b4c9ede079f866721526eba4c679ee3#diff-5f6148d0c4c01c76d240579b59a85252cc8edeed6aff5fe59dd26a41e35b893f
except removed the kv_cache interface.
Strange that if I removed the shared_head in the forward of DeepSeekMultiTokenPredictorLayer, there will be a GC error when compute_logits. I guess somewhere later there is a change that involved deepseek_mtp and logits related components.
  return self.shared_head(hidden_states)

File: vllm/model_executor/models/deepseek_v2.py (hunk 3 above)
Comment (czhu15) [reply vllm/model_executor/models/deepseek_v2.py:504]: Let me check if I can revert this interface change.

File: vllm/model_executor/models/deepseek_mtp.py (hunk 2 above)
Comment (czhu15) [reply vllm/model_executor/models/deepseek_mtp.py:91]: will create a new PR with separate commits soon.

**Comment by czhu15** [comment]
Created a new PR with separate commits for easy review and future rebase.
Pls review below PR instead.
https://github.com/HabanaAI/vllm-fork/pull/1141
Thanks.

**Comment by czhu15** [comment]
replaced with https://github.com/HabanaAI/vllm-fork/pull/1141
